queries/vhdl/highlights.scm

```diff
@@ -0,0 +1,2397 @@
+
+(selected_name
+   prefix: (selected_name
+             prefix: ((simple_name) @_p
+                      (#vim-match? @_p "^[sS][tT][dD]$"))
+             suffix: ((simple_name) @_s @error
+                      (#not-vim-match? @_s "^\%([tT][eE][xX][tT][iI][oO]|[eE][nN][vV]\)$"))
+             ) @error.region)
+; }}}
+; 13.3 Context declarations {{{
+((context_declaration
+         . [(identifier)  (extended_identifier)] @error.declaration
+   at_end: [(simple_name) (extended_simple_name)] @error .)
+           (#not-eq? @error.declaration @error))
+; }}}
+; 15.5.2 Decimal literals {{{
+((negative_exponent) @error
+ (#has-parent? @error integer_decimal based_integer))
+; }}}
+; PSL 6.1.2 Sequences {{{
+(PSL_Repeated_SERE
+   (PSL_Count
+      operator: "+"
+      (simple_expression) @error) @error.region)
+
+(PSL_Repeated_SERE
+   (PSL_Count
+      operator: "=" . "]" @error @error.missing.count) @error.region)
+
+(PSL_Repeated_SERE
+   (PSL_Count
+      operator: "->" . "]" @error @error.missing.count) @error.region)
+
+(PSL_Repeated_SERE
+   (PSL_Count
+      operator: "->"
+      (simple_expression (sign operator: "-" @error @error.expecting.positive_number)) @error.region))
+
+(PSL_Repeated_SERE
+   ((PSL_Count "[" @error @error.missing.operator) @error.region
+    (#not-has-children @error.region "*" "+" "=" "->")))
+; }}}
+; PSL 6.2.1 FL Properties {{{
+(PSL_Extended_Ocurrence_FL_Property
+   (PSL_Count
+      operator: _ @error) @error.region)
+
+(PSL_Extended_Ocurrence_FL_Property
+   operator: ["next" "next_event"]
+   (PSL_Count [
+         (ascending_range "to" @error)
+         (descending_range "downto" @error)
+         (attribute_name (predefined_designator) @error)
+   ] @error.expecting.number) @error.region)
+
+(PSL_Extended_Ocurrence_FL_Property
+   operator: ["next_a" "next_e" "next_event_a" "next_event_e"]
+   (PSL_Count
+      (simple_expression) @error @error.expecting.range) @error.region)
+
+(PSL_Extended_Ocurrence_FL_Property
+   operator: ["next_a" "next_e"]
+   (PSL_Count
+      (_
+         (simple_expression [
+            ((simple_name) @error @error.expecting.finite_range
+             (#vim-match? @error "^[iI][nN][fF]$"))
+            (sign
+               ((simple_name) @error @error.expecting.finite_range
+                (#vim-match? @error "^[iI][nN][fF]$")))
+            ]) @error.region)))
+
+(PSL_Extended_Ocurrence_FL_Property
+   operator: ["next_event"]
+   (PSL_Count
+      (simple_expression
+         (sign operator: "-" @error @error.expecting.positive_number))) @error.region)
+
+(PSL_Extended_Ocurrence_FL_Property
+   operator: ["next_event_a" "next_event_e"]
+   (PSL_Count
+      (_
+         (simple_expression
+            (sign "-" @error @error.expecting.positive_number)) @error.region)))
+; }}}
+
+(ERROR) @error @error.parser
+
+[
+   (identifier)
+   (extended_identifier)
+   (PSL_Identifier)
+   (simple_name)
+   (extended_simple_name)
+] @variable
```

Comment:
:scream:

Reply:
must have been a huge effort to port the whole standard. impressive!

Reply:
Yeah. Writing the grammar and the query took way longer than I'd admit. But I'm happy of the result. The parser with the highlight query can highlight common syntax errors very precisely.

The grammar is still missing a couple of things from the standard, such PSL OBE Expression and protected tool directives (used for encryption). I intend to implement the protected directives, but not the OBE expressions -- their syntax are not suitable for tree-sitter format, and isn't supported by many tools anyway.

There is a few other uncommon syntax errors that I intended to add to query.

Once all the standard queries (locals/tags/fold/indent) are implement I intend to extend the grammar to the latest language revision.